### PR TITLE
Local execution native interface vad

### PR DESCRIPTION
**Pull request by aitorrf**

### CODE REVIEW

File: library/src/main/java/com/voysis/sevice/local/LocalClient.kt [hunk 1]
```diff
@@ -146,9 +146,14 @@ internal class LocalClient(private val config: Config, private val vadService: P
         return Gson().toJson(response)
     }
 
-    private fun appyVAD() {
+    private fun appyVad() {
         acousticQuery.streamFuture.responseReason = VAD_RECEIVED
```
Comment (lduffy-voysis) [inline library/src/main/java/com/voysis/sevice/local/LocalClient.kt]: type? should it be applyVad?

File: library/src/main/java/com/voysis/sevice/local/LocalClient.kt [hunk 2]
```diff
@@ -146,9 +146,14 @@ internal class LocalClient(private val config: Config, private val vadService: P
...
         audioStopper.onStop()
         acousticQuery.channel.read(createBuffer()) //unblock writer loop
     }
+
+    override fun cleanUp() {
+        asrService.cleanUp()
+        vadService.cleanUp()
+    }
```
Comment (lduffy-voysis) [inline library/src/main/java/com/voysis/sevice/local/LocalClient.kt]: when should cleanup be called?  Also, can the service be used after cleanup? If not we should probably document that it cant and throw an error if the user tries to use it after cleanup.

Comment (aitorrf) [reply library/src/main/java/com/voysis/sevice/local/LocalClient.kt]: should it be called from onDestroy?

Comment (aitorrf) [reply library/src/main/java/com/voysis/sevice/local/LocalClient.kt]: I added some documentation in the service interface. As we've talked we should design a good solution for memory management at any point.

Comment (lduffy-voysis) [reply library/src/main/java/com/voysis/sevice/local/LocalClient.kt]: maybe also see if the c lib code can be optimised to reduce its memory footprint.

File: library/src/main/java/com/voysis/sevice/local/modules/VADService.kt [hunk 3]
```diff
@@ -1,45 +1,193 @@
...
+        fun get_vad_start_frame(data: Pointer): Int
+        fun get_vad_stop_frame(data: Pointer): Int
+        fun get_vad_state(data: Pointer): Int
+        fun cleanup_vad_data(data: Pointer)// USED IN ASRService
+
+        fun add_audio_to_vad_session(processor: Pointer, session: Pointer, buf: ByteBuffer, buf_size: Int, more_data: Boolean, error: IntByReference): Pointer
+    }
+
+    data class VadAudio(val data: Pointer, val audioBuffer: Pointer?, val vadAudioSize: Long, var startFrame: Int = 0, var stopFrame: Int = 0, val vadState: VADState) {
+        fun isStopTalking(): Boolean {
```
Comment (lduffy-voysis) [inline library/src/main/java/com/voysis/sevice/local/modules/VADService.kt]: better wording might be hasStoppedTalking()

File: library/src/main/java/com/voysis/sevice/local/modules/VADService.kt [hunk 4]
```diff
@@ -1,45 +1,193 @@
...
+            return vadState.equals(VADState.VT_VAD_NOT_SPEAKING) || vadState.equals(VADState.VT_VAD_END_OF_AUDIO)
+        }
+
+        fun isVadSignal(): Boolean {
```
Comment (lduffy-voysis) [inline library/src/main/java/com/voysis/sevice/local/modules/VADService.kt]: vadSignalled()

File: library/src/main/java/com/voysis/sevice/local/modules/ASRService.kt [hunk 5]
```diff
@@ -4,64 +4,142 @@ import android.util.Log
...
-internal class ASRService(private val acousticModel: PointerByReference, private val languageModel: PointerByReference, private var cLib: AcousticLibrary = Native.load(DEFAULT_NATIVE_LIBRARY_NAME, AcousticLibrary::class.java), private val nativeExecutor: ExecutorService = Executors.newSingleThreadExecutor()) : PipelineClient() {
+internal class ASRService(private val acousticModelProcessor: Pointer,
+                          private val languageModelDecoder: Pointer,
+                          private var cLib: AcousticLibrary = Native.load(DEFAULT_NATIVE_LIBRARY_NAME, AcousticLibrary::class.java),
+                          private val nativeExecutor: ExecutorService = Executors.newSingleThreadExecutor()) : PipelineClient() {
 
-    private lateinit var processor: Pointer
-    private lateinit var res: PointerByReference
+    var session: Pointer? = null
+    lateinit var streamFuture: AudioResponseFuture
```
Comment (lduffy-voysis) [inline library/src/main/java/com/voysis/sevice/local/modules/ASRService.kt]: should these be private?

File: library/src/main/java/com/voysis/sevice/local/LocalClient.kt [hunk 6]
```diff
@@ -152,7 +152,7 @@ internal class LocalClient(private val config: Config, private val vadService: P
         acousticQuery.channel.read(createBuffer()) //unblock writer loop
     }
 
-    override fun cleanUp() {
+    override fun close() {
         asrService.cleanUp()
```
Comment (oranoceallaigh) [inline library/src/main/java/com/voysis/sevice/local/LocalClient.kt]: Do the asrService and vadService _guarantee_ no exceptions will be thrown, even if cleanup cannot be completed? This is one of the places where wide-band exception handling is desirable. Close-down should be on a very-best-effort basis, but if `asrService.cleanUp()` does not complete successfully, `vadService.cleanUp()` should still have the opportunity to execute to minimise any memory leaks.

You should either do this within the cleanup methods themselves (and explicitly state in the javadocs for that method that all exceptions are handled) or you can wrap each call to cleanup at this level in a general exception handler to allow execution to continue. If you want to handle it at this layer then I'd be likely to do something like

```
try {
    asrService.cleanUp()
} finally {
    vadService.cleanUp()
}
```

That way the application will still be delivered an exception if either of the methods fails, but if `asrService` exceptions, `vadService` will still execute before the exception bubbles out to the application.

File: library/src/main/java/com/voysis/events/Exceptions.kt [hunk 7]
```diff
@@ -17,3 +17,12 @@ class PermissionDeniedException : VoysisException {
 
     constructor(message: String, throwable: Throwable) : super(message, throwable)
 }
+
+class NativeException : VoysisException {
```
Comment (oranoceallaigh) [inline library/src/main/java/com/voysis/events/Exceptions.kt]: Can you javadoc what the intended use of a `NativeException` is, please?

File: library/src/main/java/com/voysis/sevice/ServiceImpl.kt [hunk 8]
```diff
@@ -64,6 +65,15 @@ internal class ServiceImpl(private val client: Client,
         state = State.IDLE
     }
 
+    /**
+     * You can call this method to clean internal memory objects when the app is going to close
+     * These objects are necessary to run properly, so if we destroy it, you will need to create a new service
```
Comment (oranoceallaigh) [inline library/src/main/java/com/voysis/sevice/ServiceImpl.kt]: This contract needs to be documented on the `Service` interface and not here. You inherit from `Closable`, the docs on the interface class should document that closing a service invalidates it such that it cannot be re-used and a new one needs to be instantiated. You should say something like

```
Service implementations may allocate and hold operating system resources that need to be explicitly released. Closing the service will release these resources but once closed a Service cannot be reused. Note that allocating these resources may be a time intensive operation, such as model loading for local execution mode, and a calling application will need to be acutely aware of when it instantiates and discards Service implementations.
```

File: library/src/main/java/com/voysis/sevice/local/LocalClient.kt [hunk 9]
```diff
@@ -92,25 +93,24 @@ internal class LocalClient(private val config: Config, private val vadService: P
...
         }
         checkStreamStopReason(acousticQuery)
     }
 
-    private fun send(acousticQuery: AcousticQuery) {
-        notifyObservers(acousticQuery)
-        acousticQuery.record.let { it.position(it.limit()) } // make sure that record is consumed
+    private fun send(record: ByteBuffer) {
+        notifyObservers(record)
+        record.let { it.position(it.limit()) } // make sure that record is consumed
```
Comment (oranoceallaigh) [inline library/src/main/java/com/voysis/sevice/local/LocalClient.kt]: Not sure what the `let` wrapper gets you here. Seems like this line can be directly replaced with

```
record.position(record.limit())
```

Can you also avoid trailing comments, please? Put the comment on its own on the line above.

File: library/src/main/java/com/voysis/sevice/local/modules/ASRService.kt [hunk 10]
```diff
@@ -4,64 +4,142 @@ import android.util.Log
...
 
-private val DEFAULT_NATIVE_LIBRARY_NAME = "android-test"
+private val DEFAULT_NATIVE_LIBRARY_NAME = "vtasr"
 
 /**
  * ASR service decode wav to string
  *  when no moreData is detected, notify decoded String
  */
-internal class ASRService(private val acousticModel: PointerByReference, private val languageModel: PointerByReference, private var cLib: AcousticLibrary = Native.load(DEFAULT_NATIVE_LIBRARY_NAME, AcousticLibrary::class.java), private val nativeExecutor: ExecutorService = Executors.newSingleThreadExecutor()) : PipelineClient() {
+internal class ASRService(private val acousticModelProcessor: Pointer,
```
Comment (oranoceallaigh) [inline library/src/main/java/com/voysis/sevice/local/modules/ASRService.kt]: Does ktlint allow this format? All other kotlin code bases use a different intendation and bracket closing style:

```
internal class ASRService(
    private val acousticModelProcessor: Pointer,
    private val languageModelDecoder: Pointer,
    private var cLib: AcousticLibrary = Native.load(DEFAULT_NATIVE_LIBRARY_NAME, AcousticLibrary::class.java),
    private val nativeExecutor: ExecutorService = Executors.newSingleThreadExecutor()
) : PipelineClient() {    
```

If ktlint allows it then I'm not going to force a change.

File: library/src/main/java/com/voysis/sevice/local/modules/VADService.kt [hunk 11]
```diff
@@ -1,45 +1,193 @@
...
+        val resetError = IntByReference()
+        cLib.reset_vad_processor(vadProcessor, resetError)
+        if (resetError.value != 0) {
+            throw NativeException("Native Error : " + resetError.value.toString())
+        }
+    }
+
+    private fun createSession() {
+        val createSessionError = IntByReference()
+//        session = cLib.create_vad_session_default(vadProcessor, createSessionError)
```
Comment (oranoceallaigh) [inline library/src/main/java/com/voysis/sevice/local/modules/VADService.kt]: delete commented code

File: library/src/main/java/com/voysis/sevice/local/modules/VADService.kt [hunk 12]
```diff
@@ -1,45 +1,193 @@
...
-    private fun execute(acousticQuery: AcousticQuery) {
-        if (acousticQuery.streamFuture.responseReason != VAD_RECEIVED) {
-            nativeExecutor.execute {
-                vadCount++
-                Log.d("VADService", "VAD COUNT : " + vadCount)
-                if (vadCount == 50) {
-                    Log.d("VADService", "Mock VAD triggered")
-                    notifyObservers(VAD_RECEIVED)
+    private fun execute(record: ByteBuffer) {
+
```
Comment (oranoceallaigh) [inline library/src/main/java/com/voysis/sevice/local/modules/VADService.kt]: unnecessary whitespace

File: library/src/main/java/com/voysis/sevice/local/modules/VADService.kt (hunk 11 above)
Comment (aitorrf) [reply library/src/main/java/com/voysis/sevice/local/modules/VADService.kt]: I forget to mention this in yesterday meeting. We didn't decide if we want to use the default session or we are interested in to use the parameter option for any specific reason. Currently, default option is using the same values than the option with parameters.  @oranoceallaighwhat do you think?

File: library/src/main/java/com/voysis/events/Exceptions.kt (hunk 7 above)
Comment (aitorrf) [reply library/src/main/java/com/voysis/events/Exceptions.kt]: Are we interested in to differentiating the errors that come from the native code or we are ok with voysisException for everything?

File: library/src/main/java/com/voysis/sevice/local/modules/ASRService.kt (hunk 10 above)
Comment (aitorrf) [reply library/src/main/java/com/voysis/sevice/local/modules/ASRService.kt]: yes, ktlint allows it

we are using it in other projects when we have a long constructor definition.

File: library/src/main/java/com/voysis/events/Exceptions.kt (hunk 7 above)
Comment (oranoceallaigh) [reply library/src/main/java/com/voysis/events/Exceptions.kt]: I'm ok with either approach to be honest. Nothing wrong with having this specialisation but if you decide to keep it, it just needs a bit of javadoc to explain when you (as an application) will receive a `NativeException` versus any other specialisation of `VoysisException`

File: library/src/main/java/com/voysis/sevice/local/modules/VADService.kt (hunk 11 above)
Comment (oranoceallaigh) [reply library/src/main/java/com/voysis/sevice/local/modules/VADService.kt]: @ttroy50 Probably a better question for you. My opinion is that we should use the `_default` one if we're not going to allow the SDK layer to have any control over the parameters. If we're going to use the form that allows us passing the parameters then they shouldn't be hard-coded, they should come from config so that they can be tweaked at run time.

File: library/src/main/java/com/voysis/sevice/local/modules/ASRService.kt (hunk 10 above)
Comment (oranoceallaigh) [reply library/src/main/java/com/voysis/sevice/local/modules/ASRService.kt]: Ok grand

File: library/src/main/java/com/voysis/sevice/local/modules/VADService.kt (hunk 11 above)
Comment (ttroy50) [reply library/src/main/java/com/voysis/sevice/local/modules/VADService.kt]: For now default is ok as those values haven't changed in a while but we will eventually want to allow for them to come from config as they are related to the VAD model in use.
